In [ ]:
Mifkad_RAW_PATH = Path(r'C:\porat\data course final project\data\raw\mifkad 2022.csv')
df_mifkad= pd.read_csv(Mifkad_RAW_PATH).copy()

df_mifkad.head()

,LocNameHeb,LocalityCode,StatArea,StatAreaCmb,pop,pop_density,religion,sexRatio,age0_19_pcnt,age20_64_pcnt,...,size_avg,hh0_5_pcnt,hh18_24_pcnt,hh_MidatDatiyut,Computer_avg,Vehicle0_pcnt,Vehicle2up_pcnt,Parking_pcnt,own_pcnt,rent_pcnt
0,אבו גוש,472,NaN,NaN,8660,6558.4,2,114.3,33.3,60.7,...,3.4,26.4,37.5,3.0,1.5,25.0,30.3,53.3,54.7,21.4
1,אבו סנאן,472,NaN,NaN,13900,6015.2,2,108.0,32.6,59.5,...,3.5,25.2,30.8,2.0,1.7,18.0,38.9,76.4,78.4,4.9
2,אבו סנאן,473,1.0,1,5100,NaN,5,99.8,33.9,58.4,...,3.9,28.7,27.6,2.0,1.7,13.3,42.6,81.6,79.6,7.4
3,אבו סנאן,473,2.0,2,3250,NaN,2,102.0,29.8,62.2,...,3.3,19.9,33.8,2.0,1.6,19.9,36.3,69.6,75.7,5.3
4,אבו סנאן,473,3.0,3,5550,NaN,2,120.2,33.0,58.9,...,3.7,25.0,31.4,2.0,1.6,18.4,38.8,77.1,84.1,2.3


In [ ]:
df_mifkad = df_mifkad.fillna(0)
df_mifkad['StatArea'] = df_mifkad['StatArea'].astype(int)
df_mifkad['religion'] = df_mifkad['religion'].astype('category')
df_mifkad['hh_MidatDatiyut'] = df_mifkad['hh_MidatDatiyut'].astype(int)
df_mifkad.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3378 entries, 0 to 3377
Data columns (total 31 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   LocNameHeb         3378 non-null   object  
 1   LocalityCode       3378 non-null   int64   
 2   StatArea           3378 non-null   int32   
 3   StatAreaCmb        3378 non-null   object  
 4   pop                3378 non-null   int64   
 5   pop_density        3378 non-null   float64 
 6   religion           3378 non-null   category
 7   sexRatio           3378 non-null   float64 
 8   age0_19_pcnt       3378 non-null   float64 
 9   age20_64_pcnt      3378 non-null   float64 
 10  age65_pcnt         3378 non-null   float64 
 11  age_median         3378 non-null   float64 
 12  m_age_median       3378 non-null   int64   
 13  w_age_median       3378 non-null   int64   
 14  married18_34_pcnt  3378 non-null   float64 
 15  married45_54_pcnt  3378 non-null   float64 
 16  Marria

In [ ]:
df_mifkad['City_agas_code'] = df_mifkad['LocalityCode'].astype(str) + "_" + df_mifkad['StatArea'].astype(str)
df_mifkad = df_mifkad.set_index('City_agas_code')
df_mifkad = df_mifkad.drop(columns=['StatAreaCmb','LocalityCode', 'StatArea'])
df_mifkad.head(20)

,LocNameHeb,pop,pop_density,religion,sexRatio,age0_19_pcnt,age20_64_pcnt,age65_pcnt,age_median,m_age_median,...,size_avg,hh0_5_pcnt,hh18_24_pcnt,hh_MidatDatiyut,Computer_avg,Vehicle0_pcnt,Vehicle2up_pcnt,Parking_pcnt,own_pcnt,rent_pcnt
City_agas_code,,,,,,,,,,,,,,,,,,,,,
472_0,אבו גוש,8660,6558.4,2,114.3,33.3,60.7,6.0,28.0,29,...,3.4,26.4,37.5,3,1.5,25.0,30.3,53.3,54.7,21.4
472_0,אבו סנאן,13900,6015.2,2,108.0,32.6,59.5,7.9,29.0,27,...,3.5,25.2,30.8,2,1.7,18.0,38.9,76.4,78.4,4.9
473_1,אבו סנאן,5100,0.0,5,99.8,33.9,58.4,7.6,29.0,27,...,3.9,28.7,27.6,2,1.7,13.3,42.6,81.6,79.6,7.4
473_2,אבו סנאן,3250,0.0,2,102.0,29.8,62.2,8.1,29.0,28,...,3.3,19.9,33.8,2,1.6,19.9,36.3,69.6,75.7,5.3
473_3,אבו סנאן,5550,0.0,2,120.2,33.0,58.9,8.1,28.0,27,...,3.7,25.0,31.4,2,1.6,18.4,38.8,77.1,84.1,2.3
1342_0,אבו קורינאת (יישוב),2370,2366.9,2,116.1,59.5,39.6,0.9,15.0,15,...,4.9,73.3,59.3,6,1.0,35.2,2.2,49.7,67.7,2.8
1375_0,אבו תלול,4980,3075.9,2,84.5,53.8,41.6,4.6,18.0,19,...,4.1,39.8,49.0,2,1.2,42.7,4.6,61.2,92.8,0.0
652_0,אבטין,2750,7437.9,2,103.6,39.0,56.1,4.9,25.0,24,...,4.1,30.5,33.0,2,1.3,23.2,21.8,60.0,56.3,7.9
1275_0,אבטליון,470,3147.5,1,99.4,41.7,55.1,3.2,26.0,27,...,3.5,24.7,29.2,1,2.4,0.0,66.2,75.2,79.1,16.1


In [ ]:
# merge of the main dataframe with the Mifkad dataframe, on each date for the same city era the mifkad values are static

merged_df = df_loc_corona.merge(df_mifkad, on='City_agas_code', how='left')
merged_df.tail()

,town,date,accumulated_cases,accumulated_recoveries,accumulated_hospitalized,accumulated_deaths,accumulated_diagnostic_tests,accumulated_vaccination_first_dose,accumulated_vaccination_second_dose,accumulated_vaccination_third_dose,...,size_avg,hh0_5_pcnt,hh18_24_pcnt,hh_MidatDatiyut,Computer_avg,Vehicle0_pcnt,Vehicle2up_pcnt,Parking_pcnt,own_pcnt,rent_pcnt
City_agas_code,,,,,,,,,,,,,,,,,,,,,
99_0,מצפה רמון,2023-05-24,2676,2666,34,7,32481,3545,3323,2521,...,2.7,23.2,21.8,1.0,1.6,31.4,16.0,55.8,39.6,51.6
99_0,מצפה רמון,2023-05-25,2676,2666,34,7,32481,3545,3323,2521,...,2.7,23.2,21.8,1.0,1.6,31.4,16.0,55.8,39.6,51.6
99_0,מצפה רמון,2023-05-26,2676,2666,34,7,32481,3545,3323,2521,...,2.7,23.2,21.8,1.0,1.6,31.4,16.0,55.8,39.6,51.6
99_0,מצפה רמון,2023-05-27,2676,2666,34,7,32482,3545,3323,2521,...,2.7,23.2,21.8,1.0,1.6,31.4,16.0,55.8,39.6,51.6
99_0,מצפה רמון,2023-05-28,2676,2666,34,7,32482,3545,3323,2521,...,2.7,23.2,21.8,1.0,1.6,31.4,16.0,55.8,39.6,51.6


In [ ]:
merged_df.isnull().sum()

town                                        0
date                                        0
accumulated_cases                           0
accumulated_recoveries                      0
accumulated_hospitalized                    0
accumulated_deaths                          0
accumulated_diagnostic_tests                0
accumulated_vaccination_first_dose          0
accumulated_vaccination_second_dose         0
accumulated_vaccination_third_dose          0
positive_daily                              0
hospitalized_daily                          0
tested_daily                                0
infection_rate                              0
recovery_ratio                              0
recovery_index                              0
is_lockdown                                 0
LocNameHeb                             574456
pop                                    574456
pop_density                            574456
religion                               574456
sexRatio                          

In [ ]:
merged_df.shape

(1944080, 45)

In [ ]:
missing_cities = sorted(set(df_loc_corona.index.unique()) - set(df_mifkad.index.unique()))
print("Cities in df_loc_corona not found in df_mifkad:", missing_cities)

# example: for city 1020. Area 1 has no data and area 2 has data in the mifkad dataset. in the mifkad data for area 2 it mentioned as combined with area 1 (2+1) in StatAreaCmb comumn.
merged_df[merged_df.index == '1020_1'].head()

Cities in df_loc_corona not found in df_mifkad: ['1020_1', '1020_4', '1034_7', '1061_11', '1161_1', '1161_10', '1161_11', '1161_2', '1161_3', '1161_4', '1161_5', '1161_6', '1161_7', '1161_8', '1161_9', '1200_11', '1200_12', '1200_13', '1200_15', '1200_16', '1200_17', '1200_21', '1200_22', '1200_23', '1200_24', '1200_31', '1200_33', '1200_34', '1200_41', '1200_42', '1200_43', '1309_1', '1309_2', '1309_3', '1309_4', '1309_5', '1309_6', '1309_7', '168_1', '2200_2', '2200_3', '2200_4', '2200_5', '2200_6', '2200_9', '229_1', '2400_1', '2400_10', '2400_2', '2400_3', '2400_4', '2400_5', '2400_7', '2400_9', '246_1', '246_2', '246_4', '246_5', '246_6', '246_7', '246_8', '2550_1', '2600_33', '2600_34', '2600_35', '2600_36', '2610_12', '2610_13', '2610_14', '2610_15', '2610_16', '2610_21', '2610_22', '2610_24', '2610_25', '2610_26', '2610_27', '2610_31', '2610_32', '2610_33', '2610_34', '2610_35', '2610_36', '2610_37', '2610_38', '2620_1', '2620_2', '2620_3', '2620_4', '2620_5', '2620_6', '2620_8

,town,date,accumulated_cases,accumulated_recoveries,accumulated_hospitalized,accumulated_deaths,accumulated_diagnostic_tests,accumulated_vaccination_first_dose,accumulated_vaccination_second_dose,accumulated_vaccination_third_dose,...,size_avg,hh0_5_pcnt,hh18_24_pcnt,hh_MidatDatiyut,Computer_avg,Vehicle0_pcnt,Vehicle2up_pcnt,Parking_pcnt,own_pcnt,rent_pcnt
City_agas_code,,,,,,,,,,,,,,,,,,,,,
1020_1,אור עקיבא,2020-03-11,0,0,0,0,0,0,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1020_1,אור עקיבא,2020-03-12,0,0,0,0,0,0,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1020_1,אור עקיבא,2020-03-13,0,0,0,0,0,0,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1020_1,אור עקיבא,2020-03-14,0,0,0,0,7,0,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1020_1,אור עקיבא,2020-03-15,0,0,0,0,7,0,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
# since some of the data in the Mifkad 2022 was done for combined ares in a city (City_agas_code), but have values in the corona dataset (df_loc_corona), I will retrive data from one city area and apply it ao another as well.
# for example, for city 1020 there are two areas 1 and 2 in the corona dataset, data for 1 and two was done together and was recorded in area 2 in the mifkad dataset. so I will copy the values from area 2 to area 1.



neighbor_map = {
    '1020_1': '1020_2',
    '1161_11': '1161_16',
    '229_1': '229_3',
    '2600_33': '2600_32',
    '2640_11': '2640_12',
    '3000_2112': '3000_2111',
    '3000_2211': '3000_2213',
    '3000_2212': '3000_2213',
    '3000_2311': '3000_2312',
    '3000_2412': '3000_2411',
    '3000_2613': '3000_2611',
    '3000_2812': '3000_2811',
    '3000_2911': '3000_2912',
    '4000_812': '4000_813',
    '5000_115': '5000_114',
    '5000_731': '5000_732',
    '5000_734': '5000_733',
    '6600_624': '6600_623',
    '6800_12': '6800_16',
    '7200_11': '7200_12',
    '7300_34': '7300_35',
    '7700_11': '7700_12',
    '7700_15': '7700_14',
    '7700_25': '7700_26',
    '8500_11': '8500_15',
    '8600_414': '8600_415',
    '9200_1': '9200_2',
    '9500_11': '9500_12',
}


for target_city, neighbor_city in neighbor_map.items():
    mask = merged_df.index == target_city
    neighbor_values = df_mifkad.loc[neighbor_city]
    for col in df_mifkad.columns:
        merged_df.loc[mask, col] = neighbor_values[col]

merged_df[merged_df.index == '1020_1'].head()

,town,date,accumulated_cases,accumulated_recoveries,accumulated_hospitalized,accumulated_deaths,accumulated_diagnostic_tests,accumulated_vaccination_first_dose,accumulated_vaccination_second_dose,accumulated_vaccination_third_dose,...,size_avg,hh0_5_pcnt,hh18_24_pcnt,hh_MidatDatiyut,Computer_avg,Vehicle0_pcnt,Vehicle2up_pcnt,Parking_pcnt,own_pcnt,rent_pcnt
City_agas_code,,,,,,,,,,,,,,,,,,,,,
1020_1,אור עקיבא,2020-03-11,0,0,0,0,0,0,0,0,...,3.1,27.3,17.6,1.0,1.7,11.9,40.5,87.6,69.4,29.3
1020_1,אור עקיבא,2020-03-12,0,0,0,0,0,0,0,0,...,3.1,27.3,17.6,1.0,1.7,11.9,40.5,87.6,69.4,29.3
1020_1,אור עקיבא,2020-03-13,0,0,0,0,0,0,0,0,...,3.1,27.3,17.6,1.0,1.7,11.9,40.5,87.6,69.4,29.3
1020_1,אור עקיבא,2020-03-14,0,0,0,0,7,0,0,0,...,3.1,27.3,17.6,1.0,1.7,11.9,40.5,87.6,69.4,29.3
1020_1,אור עקיבא,2020-03-15,0,0,0,0,7,0,0,0,...,3.1,27.3,17.6,1.0,1.7,11.9,40.5,87.6,69.4,29.3


In [ ]:
merged_df.isnull().sum()

town                                        0
date                                        0
accumulated_cases                           0
accumulated_recoveries                      0
accumulated_hospitalized                    0
accumulated_deaths                          0
accumulated_diagnostic_tests                0
accumulated_vaccination_first_dose          0
accumulated_vaccination_second_dose         0
accumulated_vaccination_third_dose          0
positive_daily                              0
hospitalized_daily                          0
tested_daily                                0
infection_rate                              0
recovery_ratio                              0
recovery_index                              0
is_lockdown                                 0
LocNameHeb                             541584
pop                                    541584
pop_density                            541584
religion                               541584
sexRatio                          

In [ ]:
# testing missing values have been filled
merged_df.isnull().sum()

town                                        0
date                                        0
accumulated_cases                           0
accumulated_recoveries                      0
accumulated_hospitalized                    0
accumulated_deaths                          0
accumulated_diagnostic_tests                0
accumulated_vaccination_first_dose          0
accumulated_vaccination_second_dose         0
accumulated_vaccination_third_dose          0
positive_daily                              0
hospitalized_daily                          0
tested_daily                                0
infection_rate                              0
recovery_ratio                              0
recovery_index                              0
is_lockdown                                 0
LocNameHeb                             541584
pop                                    541584
pop_density                            541584
religion                               541584
sexRatio                          

In [ ]:
# Remove cities areas that are still missing Mifkad data after attempting to fill from neighbors

missing_cities = sorted(set(df_loc_corona.index.unique()) - set(df_mifkad.index.unique()))
merged_df = merged_df[~merged_df.index.isin(missing_cities)]

In [ ]:
# testing of missing values after removing missing cities areas
merged_df.isnull().sum()

town                                   0
date                                   0
accumulated_cases                      0
accumulated_recoveries                 0
accumulated_hospitalized               0
accumulated_deaths                     0
accumulated_diagnostic_tests           0
accumulated_vaccination_first_dose     0
accumulated_vaccination_second_dose    0
accumulated_vaccination_third_dose     0
positive_daily                         0
hospitalized_daily                     0
tested_daily                           0
infection_rate                         0
recovery_ratio                         0
recovery_index                         0
is_lockdown                            0
LocNameHeb                             0
pop                                    0
pop_density                            0
religion                               0
sexRatio                               0
age0_19_pcnt                           0
age20_64_pcnt                          0
age65_pcnt      

In [ ]:
# chack how many cities areas were removed
print(len(missing_cities))

#chack how many cities areas where before and after
print("Before:", df_loc_corona.index.nunique())
print("After :", merged_df.index.nunique())

490
Before: 1657
After : 1167


Please notice!
490 city areas did not have data matching in Mifkad, therefore were droped from data.

In [ ]:
print(missing_cities)

['1020_1', '1020_4', '1034_7', '1061_11', '1161_1', '1161_10', '1161_11', '1161_2', '1161_3', '1161_4', '1161_5', '1161_6', '1161_7', '1161_8', '1161_9', '1200_11', '1200_12', '1200_13', '1200_15', '1200_16', '1200_17', '1200_21', '1200_22', '1200_23', '1200_24', '1200_31', '1200_33', '1200_34', '1200_41', '1200_42', '1200_43', '1309_1', '1309_2', '1309_3', '1309_4', '1309_5', '1309_6', '1309_7', '168_1', '2200_2', '2200_3', '2200_4', '2200_5', '2200_6', '2200_9', '229_1', '2400_1', '2400_10', '2400_2', '2400_3', '2400_4', '2400_5', '2400_7', '2400_9', '246_1', '246_2', '246_4', '246_5', '246_6', '246_7', '246_8', '2550_1', '2600_33', '2600_34', '2600_35', '2600_36', '2610_12', '2610_13', '2610_14', '2610_15', '2610_16', '2610_21', '2610_22', '2610_24', '2610_25', '2610_26', '2610_27', '2610_31', '2610_32', '2610_33', '2610_34', '2610_35', '2610_36', '2610_37', '2610_38', '2620_1', '2620_2', '2620_3', '2620_4', '2620_5', '2620_6', '2620_8', '2630_26', '2640_11', '2640_2', '2640_3', '26

In [ ]:
# What are the index dtypes?
print("df_loc_corona index dtype:", df_loc_corona.index.dtype)
print("df_mifkad index dtype:", df_mifkad.index.dtype)

# How many uniques?
print("unique loc cities:", df_loc_corona.index.nunique())
print("unique mifkad cities:", df_mifkad.index.nunique())

# Sample a few keys from each to eyeball formatting
print("loc sample:", list(df_loc_corona.index.unique()[:10]))
print("mifkad sample:", list(df_mifkad.index.unique()[:10]))


df_loc_corona index dtype: object
df_mifkad index dtype: object
unique loc cities: 1657
unique mifkad cities: 3377
loc sample: ['1015_0', '1015_1', '1015_2', '1015_3', '1015_4', '1015_5', '1015_6', '1020_0', '1020_1', '1020_2']
mifkad sample: ['472_0', '473_1', '473_2', '473_3', '1342_0', '1375_0', '652_0', '1275_0', '679_0', '1115_0']


In [ ]:
from pathlib import Path
import pandas as pd

# ---------
# Config
# ---------
OUT_DIR = Path("outputs")          # change if you want
OUT_DIR.mkdir(parents=True, exist_ok=True)

# ---------
# 1) Unique keys
# ---------
loc_keys = pd.Index(df_loc_corona.index.astype(str).str.strip()).unique()
mif_keys = pd.Index(df_mifkad.index.astype(str).str.strip()).unique()

# ---------
# 2) Full-key unmatched (exact city_area, e.g., 1015_3)
# ---------
missing_in_mifkad_full = sorted(set(loc_keys) - set(mif_keys))
missing_in_loc_full    = sorted(set(mif_keys) - set(loc_keys))

df_missing_in_mifkad_full = pd.DataFrame({"city_area_key": missing_in_mifkad_full})
df_missing_in_loc_full    = pd.DataFrame({"city_area_key": missing_in_loc_full})

# Add helpful split columns for eyeballing
for df_ in (df_missing_in_mifkad_full, df_missing_in_loc_full):
    parts = df_["city_area_key"].str.split("_", n=1, expand=True)
    df_["city_base"] = parts[0]
    df_["subarea"] = parts[1].fillna("")

# Add counts per base city (so you can sort by "how bad is it")
df_missing_in_mifkad_full["n_missing_subareas_for_base"] = (
    df_missing_in_mifkad_full.groupby("city_base")["city_area_key"].transform("count")
)
df_missing_in_loc_full["n_missing_subareas_for_base"] = (
    df_missing_in_loc_full.groupby("city_base")["city_area_key"].transform("count")
)

# Save
df_missing_in_mifkad_full.sort_values(["n_missing_subareas_for_base", "city_base", "subarea"], ascending=[False, True, True]) \
    .to_csv(OUT_DIR / "missing_in_mifkad_FULLKEY.csv", index=False)

df_missing_in_loc_full.sort_values(["n_missing_subareas_for_base", "city_base", "subarea"], ascending=[False, True, True]) \
    .to_csv(OUT_DIR / "missing_in_loc_FULLKEY.csv", index=False)

print("Saved:")
print(OUT_DIR / "missing_in_mifkad_FULLKEY.csv")
print(OUT_DIR / "missing_in_loc_FULLKEY.csv")

# ---------
# (Optional but SUPER useful) 3) Base-city unmatched (e.g., 1015 only)
# ---------
loc_base = pd.Index(loc_keys.str.split("_").str[0]).unique()
mif_base = pd.Index(mif_keys.str.split("_").str[0]).unique()

missing_in_mifkad_base = sorted(set(loc_base) - set(mif_base))
missing_in_loc_base    = sorted(set(mif_base) - set(loc_base))

pd.DataFrame({"city_base": missing_in_mifkad_base}).to_csv(OUT_DIR / "missing_in_mifkad_BASE.csv", index=False)
pd.DataFrame({"city_base": missing_in_loc_base}).to_csv(OUT_DIR / "missing_in_loc_BASE.csv", index=False)

print("Also saved:")
print(OUT_DIR / "missing_in_mifkad_BASE.csv")
print(OUT_DIR / "missing_in_loc_BASE.csv")


Saved:
outputs\missing_in_mifkad_FULLKEY.csv
outputs\missing_in_loc_FULLKEY.csv
Also saved:
outputs\missing_in_mifkad_BASE.csv
outputs\missing_in_loc_BASE.csv
